core/src/main/scala/org/apache/spark/sql/rapids/tool/store/StageModel.scala

```diff
@@ -0,0 +1,210 @@
+ *
+ * @param sInfo Snapshot from the stage info loaded from the eventlog
+ */
+@Since("24.02.3")
+class StageModel private(var sInfo: StageInfo) {
+
+  val sId: Int = sInfo.stageId
+  val attemptId: Int = sInfo.attemptNumber()
+
+  @WallClock
+  @Calculated("Calculated as (submissionTime - completionTime)")
+  var duration: Option[Long] = None
+
+  // Whenever an event is triggered, the object should update the Stage info.
+  private def updatedInfo(newSInfo: StageInfo): Unit = {
+    this.sInfo = newSInfo
+    // recalculate duration
+    calculateDuration()
+  }
+
+  /**
+   * Calculate the duration of the stage.
+   * This is called automatically whenever the stage info is updated.
+   */
+  private def calculateDuration(): Unit = {
+    duration =
+      ProfileUtils.optionLongMinusOptionLong(sInfo.completionTime, sInfo.submissionTime)
+  }
+
+  def hasFailed: Boolean = {
+    sInfo.failureReason.isDefined
+  }
+
+  def getFailureReason: String = {
+    sInfo.failureReason.getOrElse("")
+  }
+
+  def getSInfoAccumIds: Iterable[Long] = {
+    sInfo.accumulables.keySet
+  }
+
+  @Calculated
+  @WallClock
+  def getDuration: Long = {
+    duration.getOrElse(0L)
+  }
+}
+
+object StageModel {
+  def apply(stageInfo: StageInfo, stageModel: Option[StageModel]): StageModel = {
+    val sModel = stageModel.getOrElse(new StageModel(stageInfo))
+    // Initialization code used to update the duration based on the StageInfo captured from Spark's
+    // eventlog
+    sModel.updatedInfo(stageInfo)
+    sModel
+  }
+}
+
+/**
+ * A class to maintain the stage models.
+ */
+@Since("24.02.3")
+class StageModelManager extends Logging {
+  // TODO: 1- Support multithreading and concurrency.
+  //       2- Convert the storage into generic that can be replaced by file-backed storage later
+  //          when needed.
+
+  // Use Nested Maps to store stageModels which should be faster to retrieve than a map of
+  // of composite key (i.e., Tuple).
+  // Composite keys would cost more because it implicitly allocates a new object every time there
+  // is a read operation from the map.
+  private val stageIdToInfo: mutable.HashMap[Int, mutable.HashMap[Int, StageModel]] =
+    new mutable.HashMap[Int, mutable.HashMap[Int, StageModel]]()
+
+  // Holds the mapping between AccumulatorIDs to Stages (1-to-N)
+  // Note that we keep it as primitive type in case a stage was not created in the original map.
+  private val accumIdToStageId: mutable.HashMap[Long, SortedSet[Int]] =
+    new mutable.HashMap[Long, SortedSet[Int]]()
+
+  /**
+   * Returns all StageModels that have been created. This includes stages with multiple attempts.
+   * @return
+   */
+  def getAllStages: Iterable[StageModel] = stageIdToInfo.values.flatMap(_.values)
+
+  /**
+   * Returns all Ids of stage objects created as a result of handling StageSubmitted/StageCompleted.
+   * In order to get all Ids that appeared within the accumulables' Info, then
```

Comment:
this comment seems a bit strange here.  accumIdToStageId is an internal datastructure not public like this function.  I would just put a comment by accumIdToStageId saying it might not contain all stages

Reply:
Removed the line causing confusion.